src/Formats/IonReader.cpp

```diff
@@ -0,0 +1,289 @@
+#include "IonReader.h"
+
+#if USE_ION
+
+#include <decNumber/decContext.h>
+#include <Common/formatIPv6.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int INCORRECT_DATA;
+extern const int ILLEGAL_COLUMN;
+}
+
+struct IonDecimalDeleter
+{
+    void operator()(ION_DECIMAL * d)
+    {
+        ion_decimal_free(d);
+        free(d);
+    }
+};
+
+struct IonIntDeleter
+{
+    void operator()(ION_INT * i)
+    {
+        ion_int_free(i);
+        free(i);
+    }
+};
+
+struct IonTimestampDeleter
+{
+    void operator()(ION_TIMESTAMP * t) { free(t); }
+};
+
+IonReader::IonReader(ReadBuffer & in_) : in(in_)
+{
+    ION_READER_OPTIONS options;
+    memset(&options, 0, sizeof(options));
+    this->current_decimal_context = {
+        MAX_DECIMAL_DIGITS, /* max digits */
+        DEC_MAX_MATH, /* max exponent */
+        -DEC_MAX_MATH, /* min exponent */
+        DEC_ROUND_HALF_EVEN, /* rounding mode */
+        DEC_Errors, /* trap conditions */
+        0, /* status flags */
+        0 /* apply exponent clamp? */
+    };
+    options.decimal_context = &(this->current_decimal_context);
+    hREADER _reader = nullptr;
+
+    auto err = ion_reader_open_buffer(
+        &_reader, reinterpret_cast<unsigned char *>(in_.buffer().begin()), static_cast<SIZE>(in_.buffer().size()), &options);
```

Comment:
We could already read something from this read buffer, so the position can be not at the beginning of the buffer.
```suggestion
    auto err = ion_reader_open_buffer(
        &_reader, reinterpret_cast<unsigned char *>(in.position()), static_cast<SIZE>(in.available()), &options);
```

Reply:
Also what happens when underlying reader reaches the end of buffer? This buffer contains only partial data and we should call `in.next()` to load the next portion of data into the buffer.  Right now it will work only when all the input data fits into the initial buffer.